worker/firewaller/firewaller.go

```diff
@@ -1262,14 +1309,42 @@ func (rd *remoteRelationData) watchLoop() error {
 		case <-rd.catacomb.Dying():
 			return rd.catacomb.ErrDying()
 		case cidrs := <-egressAddressWatcher.Changes():
-			logger.Debugf("relation ingress addresses for %v changed in model %v: %v", rd.tag, rd.remoteModelUUID, cidrs)
+			logger.Debugf("relation egress addresses for %v changed in model %v: %v", rd.tag, rd.fw.modelUUID, cidrs)
 			if err := rd.updateOfferingModel(*rd.remoteRelationId, cidrs); err != nil {
 				return errors.Trace(err)
 			}
 		}
 	}
 }
 
+func (rd *remoteRelationData) providerEndpointLoop() error {
+	logger.Debugf("starting provider endpoint loop for %v on %v ", rd.tag.Id(), rd.localApplicationTag.Id())
+	// Watch for ingress changes requested by the consuming model.
+	ingressAddressWatcher, err := rd.fw.firewallerApi.WatchIngressAddressesForRelation(rd.tag)
+	if err != nil {
+		if !params.IsCodeNotFound(err) && !params.IsCodeNotSupported(err) {
+			return errors.Trace(err)
+		}
+		logger.Infof("no ingress required for %v", rd.localApplicationTag)
+		rd.ingressRequired = false
+		return nil
+	}
+	if err := rd.fw.catacomb.Add(ingressAddressWatcher); err != nil {
```

Comment:
Why do you add this (and the egress watcher in the requirer loop) to `fw.catacomb` rather than to `rd.catacomb`? It doesn't seem like the watchers need to outlive the `remoteRelationData`, which is the only reason I can think of.